### PR TITLE
RavenDB-14451 More fixes to output to reduce collection / references of map-reduce index

### PR DESCRIPTION
**Pull request by arekpalinski**

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs [hunk 1]
```diff
@@ -400,54 +400,41 @@ public void Execute(DocumentsOperationContext context, Transaction indexWriteTra
...
-                            if (existingReferenceDocument.Data.TryGet(nameof(OutputReduceToCollectionReference.ReduceOutputs), out BlittableJsonReaderArray ids) == false)
-                                ThrowIdsPropertyNotFound(existingReferenceDocument.Id);
-
-                            if (ids.Modifications == null)
-                                ids.Modifications = new DynamicJsonArray();
 
-                            foreach (var reduceOutputId in referencesOfReduceOutput.Value)
+                        var referenceDoc = new DynamicJsonValue
+                        {
+                            [nameof(OutputReduceToCollectionReference.ReduceOutputs)] = new DynamicJsonArray(referencesOfReduceOutput.Value),
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs]: It should be: `uniqueIds` here

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs]: and need a test for this

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs]: Thanks for catching this. Regarding a test - I saw it on your end but I'm not actually sure how it  happened.

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs]: Fixed in https://github.com/ravendb/ravendb/pull/10055/commits/017ca971f6456ff43e7a30c71c982c7203309dd7

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs]: Test added in https://github.com/ravendb/ravendb/pull/10055/commits/15285e3e932b39e3fbd946928b19205f74ca34ad. It was happening on index reset

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/DeleteReduceOutputDocumentsCommand.cs [hunk 2]
```diff
@@ -117,7 +122,13 @@ private void CleanupReferenceDocuments(DocumentsOperationContext context, Dictio
 
                     using (var doc = context.ReadObject(referenceDocument.Data, referenceDocument.Id))
                     {
-                        _database.DocumentsStorage.Put(context, referenceDocument.Id, null, doc);
+                        if (doc.TryGet(nameof(OutputReduceToCollectionReference.ReduceOutputs), out BlittableJsonReaderArray updatedIds) == false)
+                            ThrowIdsPropertyNotFound(referenceDocument.Id);
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/DeleteReduceOutputDocumentsCommand.cs:126]:
```suggestion
                            ThrowReduceOutputsPropertyNotFound(referenceDocument.Id);
```

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs [hunk 3]
```diff
@@ -400,54 +400,41 @@ public void Execute(DocumentsOperationContext context, Transaction indexWriteTra
...
-                                [nameof(OutputReduceToCollectionReference.ReduceOutputs)] = new DynamicJsonArray(referencesOfReduceOutput.Value),
-                                [Constants.Documents.Metadata.Key] = new DynamicJsonValue
-                                {
-                                    [Constants.Documents.Metadata.Collection] = $"{_outputReduceToCollection}/References"
-                                }
-                            };
-
-                            using (var referenceJson = context.ReadObject(referenceDoc, "reference-of-reduce-output", BlittableJsonDocumentBuilder.UsageMode.ToDisk))
+                            if (existingReferenceDocument.Data.TryGet(nameof(OutputReduceToCollectionReference.ReduceOutputs), out BlittableJsonReaderArray existingIds) == false)
+                                ThrowIdsPropertyNotFound(existingReferenceDocument.Id);
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs]:
```suggestion
                                ThrowReduceOutputsPropertyNotFound(existingReferenceDocument.Id);
```

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionActions.cs [hunk 4]
```diff
@@ -114,57 +119,44 @@ public OutputReduceToCollectionCommand CreateCommand(JsonOperationContext indexC
...
             return _prefixesOfReduceOutputDocumentsToDelete;
         }
 
-        public bool HasDocumentsToDelete()
-        {
-            return _prefixesOfReduceOutputDocumentsToDelete != null && _prefixesOfReduceOutputDocumentsToDelete.Count > 0;
-        }
-
         public bool HasDocumentsToDelete(TransactionOperationContext indexContext)
         {
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionActions.cs:155]: Can we rely on the value of `_prefixesOfReduceOutputDocumentsToDelete` here?
(this is now updated after the transaction was commited)

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/DeleteReduceOutputDocumentsCommand.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/DeleteReduceOutputDocumentsCommand.cs:126]: Renamed in https://github.com/ravendb/ravendb/pull/10055/commits/5902446451ad63e2740e270632e414237edef4e5

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs (hunk 3 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs]: Renamed in https://github.com/ravendb/ravendb/pull/10055/commits/5902446451ad63e2740e270632e414237edef4e5

File: src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionActions.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionActions.cs:155]: For the sake of safety I think we're better off by relying on values which are persistent. We pass `indexContext` with open transaction so it seems to be natural to be sure that the value is as of current transaction, not the in-memory value which could be updated meanwhile (even if it's not the case in current impl).

**Review by grisha-kotler — APPROVED** [review]

**Pull request merged**